Template MathFunctions on Dimension and Frame

geoffrey4444 opened the pull request:
## Proposed changes
This PR generalizes MathFunctions to support higher-dimension MathFunctions that are templated on DataType and Dimension. The MathFunctions continue to work for both doubles and DataVectors.

The existing 1D interface has the same MathFunction accept and return either doubles or DataVectors and hard-codes Frame::Inertial. The new interface, which works for higher dimensions, templates on Dimension and Frame and overloads on double/DataVector.

In this implementation, the 1D case is treated specially, to preserve the existing interface, except that users now must specify the data type, dimension, and frame.

Needed for specifying constraint damping scalars using MathFunctions and input-file options, instead of hard-coding them.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

geoffrey4444 committed:
Template MathFunctions on dimension and frame

geoffrey4444 commented on src/PointwiseFunctions/MathFunctions/MathFunction.hpp:
```diff
@@ -23,13 +23,51 @@ class Sinusoid;
  */
-template <size_t VolumeDim>
+template <size_t VolumeDim, typename T = tmpl::list<>,
+          typename Fr = Frame::Inertial>
```
Delete default for T

geoffrey4444 replied:
D'oh! Didn't factor my commits correctly...this is already done

geoffrey4444 commented on src/PointwiseFunctions/MathFunctions/MathFunction.hpp:
```diff
@@ -23,13 +23,51 @@ class Sinusoid;
  * Partial template specialization of MathFunction which encodes a
- * function \f$R \to R\f$.
+ * function \f$R \to R\f$, where the input `x` is either a `double` or a
+ * `DataVector`.
```
Would be good for 1D case to also accept tensors

geoffrey4444 commented on src/PointwiseFunctions/MathFunctions/Gaussian.hpp:
```diff
@@ -72,24 +73,111 @@ class Gaussian<1, T, Frame::Inertial>
+  Gaussian(Gaussian&& /*rhs*/) noexcept = default;
+  Gaussian& operator=(Gaussian&& /*rhs*/) noexcept = default;
+
+  Scalar<T> operator()(const tnsr::I<T, VolumeDim, Fr>& x) const
```
Don't template class on type, make a private impl function that's templated, and make overloaded public functions that call that private impl function

geoffrey4444 commented on src/PointwiseFunctions/MathFunctions/Gaussian.hpp:
```diff
@@ -72,24 +73,111 @@ class Gaussian<1, T, Frame::Inertial>
+
+template <size_t VolumeDim, typename T, typename Fr>
+bool operator!=(const Gaussian<VolumeDim, T, Fr>& lhs,
+                const Gaussian<VolumeDim, T, Fr>& rhs) noexcept;
```
Make == a friend function, no accessor functions

`friend bool operator==(const Gaussian& lhs, const Gaussian& rhs) { return ... }`

geoffrey4444 replied:
I did this, but for !=, I had trouble getting the code to compile and link unless I also made != a friend function. Otherwise, I got errors about undefined or multiply defined symbols...if it's not ok for != to be a friend function, could someone please give me a solution for the alternative?

geoffrey4444 commented on src/PointwiseFunctions/MathFunctions/MathFunction.hpp:
```diff
@@ -53,15 +54,15 @@ class MathFunction : public PUP::able {
 
   /// Returns the third partial derivatives of the function at 'x'.
-  virtual tnsr::ijj<T, VolumeDim> third_deriv(
+  virtual tnsr::iii<T, VolumeDim, Fr> third_deriv(
```
Have the 1D tesnor interface just call the DataVector interface

geoffrey4444 commented on tests/Unit/PointwiseFunctions/MathFunctions/Test_Gaussian.cpp:
```diff
@@ -83,6 +88,91 @@ void test_data_vector(const double amplitude, const double width,
+    gsl::at(center, i) = real_dis(gen);
+  }
+
+  const auto nn_generator = make_not_null(&gen);
```
check with random values & python

see equation of state tests...  TestHelpers::EquationsOfState::check()

if constexpr (bool) {
  // only compiles if bool == true
}

geoffrey4444 commented on tests/Unit/PointwiseFunctions/MathFunctions/Test_Gaussian.cpp:
```diff
@@ -5,84 +5,101 @@
+
+  pypp::SetupLocalPythonEnvironment{"PointwiseFunctions/MathFunctions/Python"};
+
+  using VolumeDims = brigand::integral_list<size_t, 1, 2, 3>;
```
In next fixup commit, `brigand::integral_list` -> `tmpl::integral_list`

geoffrey4444 commented on tests/Unit/Helpers/PointwiseFunctions/MathFunctions/TestHelpers.hpp:
```diff
@@ -0,0 +1,164 @@
+ * The python functions must be added to
+ * tests/Unit/PointwiseFunctions/MathFunctions/Python/TestFunctions.py. The
+ * prefix for each class of MathFunction is arbitrary, but should generally
+ * be descriptive (e.g. 'gaussian', 'sinusoid', 'powx') of the MathFunction.
```
In next fixup commit, `powx` -> `pow_x` to match `TestFunctions.py`

geoffrey4444 commented:
OK, this is now ready for a look, @kidder @nilsdeppe! 

Thanks for your design advice and technical help! The MathFunctions now template on frame and dimension and overload on data type. I also overhauled all the MathFunction tests to use a new test helper that calls python functions, instead of coding again in C++. In the 1D case, the test additionally checks that the old double/data vector interface agrees with the new tensor interface

geoffrey4444 commented on src/PointwiseFunctions/MathFunctions/Gaussian.hpp:
```diff
@@ -62,21 +71,38 @@ class Gaussian : public MathFunction<1> {
   DataVector third_deriv(const DataVector& x) const noexcept override;
+  using MathFunction<1, Fr>::third_deriv;
+
+  double amplitude() const noexcept { return amplitude_; }
```
@geoffrey4444 Note to self: in next fixup commit, remove these accessor functions that I forgot to remove

kidder replied:
ping?

kidder commented on src/PointwiseFunctions/MathFunctions/Gaussian.hpp:
```diff
@@ -47,11 +53,14 @@ class Gaussian : public MathFunction<1> {
   explicit Gaussian(CkMigrateMessage* /*unused*/) noexcept {}
 
   Gaussian(double amplitude, double width, double center) noexcept;
+  Gaussian(double amplitude, double width, const std::array<double, 1>& center)
```
noexcept and move into cpp file

kidder commented on src/PointwiseFunctions/MathFunctions/Gaussian.hpp:
```diff
@@ -85,12 +111,123 @@ class Gaussian : public MathFunction<1> {
+};
+
+/*!
+ *  \ingroup MathFunctionsGroup
```
indent by only one blank space with respect to the `*`

kidder commented on src/PointwiseFunctions/MathFunctions/Gaussian.hpp:
```diff
@@ -85,12 +111,123 @@ class Gaussian : public MathFunction<1> {
+
+/*!
+ *  \ingroup MathFunctionsGroup
+ *  \brief 1D Gaussian \f$f = A \exp\left(-\frac{(x-x_0)^2}{w^2}\right)\f$
```
this isn't 1D

geoffrey4444 replied:
It might or might not be, so I just removed the "1D"

kidder commented on src/PointwiseFunctions/MathFunctions/Gaussian.hpp:
```diff
@@ -85,12 +111,123 @@ class Gaussian : public MathFunction<1> {
+  using options = tmpl::list<Amplitude, Width, Center>;
+
+  static constexpr OptionString help = {
+      "Applies a Gaussian function to the input coordinates"};
```
Computes a Gaussian about an arbitrary coordinate center with given width and amplitude

kidder commented on src/PointwiseFunctions/MathFunctions/Gaussian.hpp:
```diff
@@ -47,11 +53,14 @@ class Gaussian : public MathFunction<1> {
   static constexpr OptionString help = {
       "Applies a Gaussian function to the input value"};
```
see the suggested wording below

nilsdeppe replied:
ping

kidder commented on src/PointwiseFunctions/MathFunctions/MathFunction.hpp:
```diff
@@ -6,37 +6,91 @@
 
+/*!
+ *  \ingroup MathFunctionsGroup
+ *  Encodes a function \f$R^n \to R\f$ where n is `VolumeDim` and where the
```
again only one space after the `*`

nilsdeppe replied:
ping

kidder commented on src/PointwiseFunctions/MathFunctions/MathFunction.hpp:
```diff
@@ -6,37 +6,91 @@
+  ~MathFunction() override = default;
+
+  /// Returns the value of the function at the coordinate 'x'.
+  virtual Scalar<double> operator()(
```
make these group doxygen comments with the `// @{` and `\\ @}`

kidder commented on src/PointwiseFunctions/MathFunctions/MathFunction.hpp:
```diff
@@ -47,29 +101,64 @@ class MathFunction<1> : public PUP::able {
   MathFunction& operator=(MathFunction&& /*rhs*/) noexcept = default;
   ~MathFunction() override = default;
 
-  //@{
```
don't remove these, make them all this way

kidder commented on tests/Unit/PointwiseFunctions/MathFunctions/Test_Sinusoid.cpp:
```diff
@@ -5,68 +5,75 @@
-  CHECK(sinusoid != sinusoid2);
+SPECTRE_TEST_CASE("Unit.PointwiseFunctions.MathFunctions.Sinusoid",
+                  "[PointwiseFunctions][Unit]") {
+  const DataVector dv{{std::numeric_limits<double>::signaling_NaN(),
```
`dv{5}`

kidder commented on tests/Unit/PointwiseFunctions/MathFunctions/Test_PowX.cpp:
```diff
@@ -6,101 +6,63 @@
-  }
+SPECTRE_TEST_CASE("Unit.PointwiseFunctions.MathFunctions.PowX",
+                  "[PointwiseFunctions][Unit]") {
+  const DataVector dv{{std::numeric_limits<double>::signaling_NaN(),
```
`dv{5}`

kidder commented on tests/Unit/PointwiseFunctions/MathFunctions/Test_Gaussian.cpp:
```diff
@@ -5,84 +5,101 @@
+
+SPECTRE_TEST_CASE("Unit.PointwiseFunctions.MathFunctions.Gaussian",
+                  "[PointwiseFunctions][Unit]") {
+  const DataVector dv{{std::numeric_limits<double>::signaling_NaN(),
```
`dv{5}`

kidder reviewed (changes requested):
reviewed with @nilsdeppe

geoffrey4444 commented:
Sorry for the delay responding to your review! It's ready for another look @nilsdeppe @kidder

nilsdeppe commented:
@geoffrey4444 just a couple pings. LGTM. Once you push the fixup for the last couple things I'll get @kidder to look at it

geoffrey4444 commented:
@kidder @nilsdeppe thanks for the comments! I think I addressed them all and squashed, and the tests pass, so it's ready for you to take another look when you get a chance
